### PR TITLE
Replace lazy MFSDP context initialization with an explicit scope

### LINKED ISSUE #5616: MFSDP v2 should support HSDP and HFSDP placements
MFSDP v2 currently rejects the placement differences needed for HSDP/HFSDP.

`megatron_fsdp/experimental/parameter_group.py` requires `main_grad` and `main_weight` to have identical placements and raises until HSDP/HFSDP support is implemented. That blocks hybrid sharding modes where weights, gradients, and optimizer state intentionally use different inner-DP / outer-DP layouts.

Implementation detail:

- First microbatch: all-gather across DP-outer.
- Last microbatch: reduce-scatter or all-reduce across DP-outer.
- DP-outer communication should probably run on a separate stream so it can overlap with DP-inner communication.

Expected outcome:

- Define MFSDP v2 layout semantics for HSDP and HFSDP.
- Support HSDP outer-DP replication / gradient synchronization.
- Support HFSDP optimizer-state sharding across inner and outer DP ranks.
- Add focused tests for the required placement and collective transitions.

Development PRs:
- #5652

### PR DESCRIPTION
**Pull request by wujingyue**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

## Why

MFSDP v2 has lazily initialized its runtime context on the first forward to stay close to PyTorch FSDP2. That defers stream creation, root discovery, module naming, and prefetch-order construction until runtime, adding lifecycle checks and coupling context setup to forward hooks.

[#6187](https://github.com/NVIDIA/Megatron-LM/pull/6187) is a concrete example of the resulting complexity. Because the reduce-scatter stream is unavailable when `fully_shard()` constructs parameter groups, `main_grad` must be materialized lazily so its allocation occurs on the stream that later releases it. With an eagerly initialized context, stream-dependent state can instead be constructed against the intended stream up front.

## What changes

This PR introduces an explicit construction scope:

```python
with fully_shard_context(device=device) as context:
    fully_shard(module_a, mesh=mesh, placements=placements)
    fully_shard(module_b, mesh=mesh, placements=placements)
```

Every `fully_shard()` call must run inside `fully_shard_context()`. Calling it without an active context raises a hard error. Nested construction scopes are also rejected.

The context eagerly creates its communication streams. Enclosed FSDP modules naturally inherit that context and register with it. When the scope exits, the context discovers roots, assigns module names, and builds the static forward and backward prefetch orders. Forward execution before scope exit also raises a hard error.

`microbatch()` now takes the context directly, so last-microbatch state is context-scoped rather than discovered by walking from a root module.

## Other benefits

1. All FSDP modules constructed under the same `fully_shard_context()` participate in one cross-root overlap schedule. Users no longer need an artificial `fully_shard(root_model)` call merely to make independently sharded children share streams and overlap communication.
2. The explicit context provides a natural home for context-wide configuration and state, including flags such as symmetric-memory enablement.

### CODE REVIEW

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py [hunk 1]
```diff
@@ -28,36 +28,76 @@
...
+        visited: set[FsdpModule] = set()
+        root_modules: list[FsdpModule] = []
+        for module in reversed(self._registered_modules):
+            if module in visited:
+                continue
+            root_modules.append(module)
+            _collect_fsdp_modules_under(cast(nn.Module, module), visited)
+        root_modules.reverse()
+        for root_module in root_modules:
+            root_module._is_root = True
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py]: Note to myself: there's an easy way to compute root_modules without assuming the order of registered_modules

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py [hunk 2]
```diff
@@ -17,14 +17,48 @@
...
 
 from ..mixed_precision import MixedPrecisionPolicy
 from .module import FsdpContext, FsdpModule
 from .placement import MeshAxis, Placements
 
+_FSDP_CONTEXT = ContextVar[FsdpContext | None]("megatron_fsdp_context", default=None)
+
+
+@contextmanager
+def fully_shard_context(device: torch.device | None = None) -> Iterator[FsdpContext]:
```
Comment (Autumn1998) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py:34]: Example of silent misbehavior:

```
with fully_shard_context(device=device) as ctx1:
    fully_shard(model.inner, mesh=mesh, placements=placements)
with fully_shard_context(device=device) as ctx2:
    fully_shard(model, mesh=mesh, placements=placements)
# ctx2.finalize() silently includes model.inner in its forward_order,
# while model.inner._is_root=True from ctx1 still holds.

```
Maybe worth adding a guard or UT that detects when an FsdpModule being registered already belongs to another context, and raises explicitly.

Also, If we require all `fully_shard` calls to live under a single continuous fully_shard_context, does that make it harder to align with the FSDP2 API? (I recall the plan was to achieve compatibility via a wrapper)

File: tests/unit_tests/distributed/mfsdp_v2/test_fully_shard.py [hunk 3]
```diff
@@ -238,7 +239,7 @@ def train(model, optimizer, log_prefix) -> list[torch.Tensor]:
 
             for microbatch_index, (microbatch_x, microbatch_target) in enumerate(microbatches):
                 is_last = microbatch_index == num_microbatches - 1
-                with microbatch(model, is_last=is_last):
+                with microbatch(context, is_last=is_last):
```
Comment (Autumn1998) [inline tests/unit_tests/distributed/mfsdp_v2/test_fully_shard.py:268]: If the context is already attached to the model, would it be cleaner to retrieve it via something like model.get_context()?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py (hunk 2 above)
Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py:34]: > Maybe worth adding a guard or UT that detects when an FsdpModule being registered already belongs to another context, and raises explicitly.

Yep! Good idea.

File: tests/unit_tests/distributed/mfsdp_v2/test_fully_shard.py (hunk 3 above)
Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_fully_shard.py:268]: If `model` is an `FsdpModule`, it already has `model.context`.

Otherwise, I’d prefer fewer places to track global state. Since this PR introduces `fully_shard_context`, we can reuse it for all knobs shared across `FsdpModule`s. Along those lines, perhaps `fully_shard_optimizer` should take `context` rather than `model` as well.

**Review by shjwudp — APPROVED** [review]

**Review by Phlip79 — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]
